### PR TITLE
Enable user to configure modal's scrollBehavior

### LINKED ISSUE #2327: Add ability to configure modal scrollbar position
### Please confirm that you have searched existing issues in the repo

Yes, I have searched the existing issues

### Any related issues?

#2322

### What is the area that this feature belongs to?

Other

### Is your feature request related to a problem? Please describe.

Currently, we experience a problem of unexpectedly closing the modal window when clicking on the scrollbar (if the scrollbar is outside of the modal). PR #2322 moves the scrollbar into the modal window, hence resolving the issue. However, we should explore supporting both modes if possible.

### Describe the solution you'd like

https://github.com/MarkBind/markbind/pull/2322#issuecomment-1632737944

### Describe alternatives you've considered

This might become a non-issue if we migrate to Vue 3 and upgrade our vue-final-modal version to latest.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jmestxr**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Closes #2327
Fixes #2322

**Overview of changes:**
User can now configure scroll position of modal by setting `scroll-behavior` prop to be `'inside'` (default) or `'outside'`, e.g. `<modal scroll-behavior='outside'> ...`

The modal CSS styles for the respective `scroll-behavior` values are set within the `opened()` method which is called after modal is opened (see @opened event [here](https://v3.vue-final-modal.org/guide/events/#opened)).

For `scroll-behavior = 'outside'`, basically what I did was:

1. Set the `overflow-y` property of the outermost modal container to `scroll`, which allowed user to mouseclick on scrollbar without causing modal to close unexpectedly. However, it resulted in this:

<br>

Screenshot 2023-08-17 at 12 09 28 AM

<br><br>

where the grey area is the part that user can click to close the modal and the white area is the overflow part.

<br><br>

2. To solve the above issue, I used javascript to query the modal content's total height (including the overflow part) to adjust the height of the grey part to fill up the white area, resulting in this:

<br>

Screenshot 2023-08-17 at 12 12 13 AM

<br><br>

3. Then adjust the background colors of relevant containers to get the following:

https://github.com/MarkBind/markbind/assets/87931905/ec0a9d91-9b86-47f8-890b-6e058e7d47f3

**Anything you'd like to highlight/discuss:**

When both `scroll-behavior='outside'` and `center='true'` are set, modal does not render properly when content overflows. To fix this, I modified the code such that `center='true'` only takes effect when `scroll-behavior='inside'`.

**Testing instructions:**

1. cd into `packages/cli/test/functional/test_site` and run `markbind serve -d`
1. navigate to `test_site/testModals.html` in browser and test out the modals

**Proposed commit message: (wrap lines at 72 characters)**

---

**Checklist:** :ballot_box_with_check:

- [x] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [ ] Linked all related issues
- [ ] No unrelated changes

### CODE REVIEW

**Comment by damithc** [comment]
@jmestxr is this PR ready for review?

**Comment by jmestxr** [comment]
> @jmestxr is this PR ready for review?

@damithc hi prof, this PR is ready for review

**Comment by damithc** [comment]
@MarkBind/active-devs for your review ...

File: packages/vue-components/src/Modal.vue [hunk 1]
```diff
@@ -109,16 +114,39 @@ export default {
...
+    },
   },
   methods: {
     close() {
       this.show = false;
     },
+    opened() {
+      // To adjust modal styles according to scrollBehavior
+
+      const modal = $vfm.get(this.id)[0].$el;
```
Comment (tlylt) [inline packages/vue-components/src/Modal.vue:134]: Do the properties mentioned [here](https://v2.vue-final-modal.org/guide/properties/) help you in the stuff you are doing below?

File: packages/vue-components/src/Modal.vue [hunk 2]
```diff
@@ -109,16 +114,39 @@ export default {
...
+      const modalContainer = modal.querySelector('.vfm__container');
```
[the remark this answers is not in the document]
Comment (jmestxr) [reply packages/vue-components/src/Modal.vue]: Ah, I can use those to refactor some of the styles to the <style> sheet, together with data() props. But I still need to use javascript to set `overflow-y` to the outermost container, and query the overflowed content height to adjust the modal container height.

I have just committed those changes

**Comment by jmestxr** [comment]
> Wondering whether bootstrap Modal's`modal-dialog-scrollable` will be helpful here? https://getbootstrap.com/docs/5.1/components/modal/#scrolling-long-content

Ah, this actually solves a previous issue in a more straightforward manner: https://github.com/MarkBind/markbind/pull/2322

File: packages/vue-components/src/Modal.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlylt) [reply packages/vue-components/src/Modal.vue]: For outermost container, does using <vue-final-modal> class="abc" work?

[the remark this answers is not in the document]
Comment (jmestxr) [reply packages/vue-components/src/Modal.vue]: It works. Thanks!

**Comment by tlylt** [comment]
@jmestxr 
Followed your testing step here:
> Testing instructions:
cd into packages/cli/test/functional/test_site and run markbind serve -d
navigate to test_site/testModals.html in browser and test out the modals

When clicking on the outside scrollbar, i'm getting the same old error of auto closing the entire modal. Could you verify/re-check?

Here's the gif:
4JKApwyuNw

In fact I just checked the preview of UG [here](https://deploy-preview-2355--markbind-master.netlify.app/userguide/components/popups#modals:~:text=This%20is%20a%20trigger%20for%20a%20modal%20with%20scrollbar%20positioned%20outside..)

Encountering this exact error as well.

**Comment by jmestxr** [comment]
> @jmestxr Followed your testing step here:
> 
> > Testing instructions:
> > cd into packages/cli/test/functional/test_site and run markbind serve -d
> > navigate to test_site/testModals.html in browser and test out the modals
> 
> When clicking on the outside scrollbar, i'm getting the same old error of auto closing the entire modal. Could you verify/re-check?
> 
> Here's the gif: 4JKApwyuNw
> 
> In fact I just checked the preview of UG [here](https://deploy-preview-2355--markbind-master.netlify.app/userguide/components/popups#modals:~:text=This%20is%20a%20trigger%20for%20a%20modal%20with%20scrollbar%20positioned%20outside..)
> 
> Encountering this exact error as well.

Apologies! I have just done a fix, notably for the calculation of `isContentOverflow`. The behavior should be correct now

**Comment by tlylt** [comment]
> Apologies! I have just done a fix, notably for the calculation of `isContentOverflow`. The behavior should be correct now

Just tried it, seems very janky?
- sometimes it just doesn't work, when i click on the scrollbar it closes the entire modal

- if i resize the window, it behaviors weirdly
image

**Comment by jmestxr** [comment]
> > Apologies! I have just done a fix, notably for the calculation of `isContentOverflow`. The behavior should be correct now
> 
> Just tried it, seems very janky?
> 
> * sometimes it just doesn't work, when i click on the scrollbar it closes the entire modal
> * if i resize the window, it behaviors weirdly

Okay, perhaps dynamically changing the height isn't the best way to go. Solution will become too unnecessarily complex just for this issue. Either way it doesn't solve the actual problem that the scroll-bar is within the closable area of the modal. Perhaps should instead try migrating vue-final-modal to higher versions which doesn't seem to have this issue: https://vue-final-modal.org/use-cases/modal-long-scroll

I think I'll close this PR and meanwhile open a new issue for solving #2322 using `modal-dialog-scrollable`?

**Comment by tlylt** [comment]
> Okay, perhaps dynamically changing the height isn't the best way to go. Solution will become too unnecessarily complex just for this issue. Either way it doesn't solve the actual problem that the scroll-bar is within the closable area of the modal. Perhaps should instead try migrating vue-final-modal to higher versions which doesn't seem to have this issue: [vue-final-modal.org/use-cases/modal-long-scroll](https://vue-final-modal.org/use-cases/modal-long-scroll)

Agreed. Good exploration still 👍 I think the solution that I would go with is to either wait till we migrate vue-final-modal, or do a deep dive into the package, and find out what changes did they make to fix this issue in newer versions (might take time and have complicated issues, so I'm not recommending).

> I think I'll close this PR and meanwhile open a new issue for solving #2322 using `modal-dialog-scrollable`?

Sure, if it turns out that using `modal-dialog-scrollable` can help significantly improve the implementation over the existing one then please go ahead. If not we can keep it as is.

**Comment by jmestxr** [comment]
> > Okay, perhaps dynamically changing the height isn't the best way to go. Solution will become too unnecessarily complex just for this issue. Either way it doesn't solve the actual problem that the scroll-bar is within the closable area of the modal. Perhaps should instead try migrating vue-final-modal to higher versions which doesn't seem to have this issue: [vue-final-modal.org/use-cases/modal-long-scroll](https://vue-final-modal.org/use-cases/modal-long-scroll)
> 
> Agreed. Good exploration still 👍 I think the solution that I would go with is to either wait till we migrate vue-final-modal, or do a deep dive into the package, and find out what changes did they make to fix this issue in newer versions (might take time and have complicated issues, so I'm not recommending).
> 
> > I think I'll close this PR and meanwhile open a new issue for solving #2322 using `modal-dialog-scrollable`?
> 
> Sure, if it turns out that using `modal-dialog-scrollable` can help significantly improve the implementation over the existing one then please go ahead. If not we can keep it as is.

Will do. Thanks for the review @tlylt !

**Pull request closed without merging**